File: EMS/Network/EMSRoomsParser.m
//
//  EMSRoomsParser.m
//

#import "EMS-Swift.h"

#import "EMSRoomsParser.h"

#import "CJCollection.h"
#import "CJItem.h"

static const DDLogLevel ddLogLevel = DDLogLevelDebug;

@implementation EMSRoomsParser

- (NSArray *)processData:(NSData *)data forHref:(NSURL *)href error:(NSError **)error {
    NSError *parseError = nil;

    CJCollection *collection = [CJCollection collectionForNSData:data error:&parseError];

    if (!collection) {
        DDLogError(@"Failed to retrieve rooms %@ - %@ - %@", href, parseError, [parseError userInfo]);

        *error = parseError;

        return [NSArray array];
    }

    NSMutableArray *temp = [[NSMutableArray alloc] init];

    [collection.items enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        CJItem *item = (CJItem *) obj;

        EMSRoom *room = [[EMSRoom alloc] init];

        room.href = item.href;

        [item.data enumerateObjectsUsingBlock:^(id dataObj, NSUInteger dataIdx, BOOL *dataStop) {
            NSDictionary *dict = (NSDictionary *) dataObj;

            NSString *field = dict[@"name"];
            NSString *value = dict[@"value"];

            if ([@"name" isEqualToString:field]) {
                room.name = value;
            }
        }];

        [temp addObject:room];
    }];

    return [NSArray arrayWithArray:temp];
}

- (void)parseData:(NSData *)data forHref:(NSURL *)href {
    NSError *error;

    NSArray *collection = [self processData:data forHref:href error:&error];

    [self.delegate finishedRooms:collection forHref:href error:error];
}

@end

File: EMS/Network/Parsers/EMSRoomsParser.swift
import Foundation
import CocoaLumberjack

@objc protocol EMSRoomsParserDelegate {
    optional func finishedRooms(rooms: [EMSRoom], forHref href:NSURL, error:NSError?)
}

@objc class EMSRoomsParser : NSObject {
    var delegate : EMSRoomsParserDelegate?
    
    func processData(data: NSData, forHref href:NSURL) throws -> [EMSRoom] {
        
        do {
            let collection = try CJCollection(forNSData:data)
            
            return collection.items.map({ item in
                let room = EMSRoom()
                
                room.href = item.href
                
                for datum in item.data {
                    if let field = datum["name"] as? String {
                        if let value = datum["value"] as? String {
                            switch (field) {
                            case "name":
                                room.name = value
                            default:
                                DDLogDebug("Ignored string field \(field) for room")
                            }
                        }
                    }
                }
                
                return room
            })
        } catch let error as NSError {
            DDLogError("Failed to retrieve rooms \(href) - \(error)")
            
            throw error
        }
    }
    
    func parseData(data: NSData, forHref url: NSURL) {
        do {
            let collection = try self.processData(data, forHref: url)
            
            self.delegate?.finishedRooms?(collection, forHref: url, error: nil)
        } catch let error as NSError {
            self.delegate?.finishedRooms?([], forHref: url, error: error)
        }
    }
}
